### PR TITLE
[Translator] move loading catalogue out of Translator.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Fixed tickets | ~ |
| Tests pass? | yes |
| License | MIT |
#### Before

``` php
$translator = new Translator('fr');
$translator->addLoader('array', new ArrayLoader());
$translator->addResource('array', array('Hello World!' => 'Bonjour'), 'fr');

echo $translator->trans('Hello World!')."\n";
```
#### After

``` php
$resourceCatalogue = new ResourceMessageCatalogueProvider();
$resourceCatalogue->addLoader('array', new ArrayLoader());
$resourceCatalogue->addResource('array', array('Hello World!' => 'Bonjour'), 'fr');
$translator = new Translator('fr', $resourceCatalogue);

echo $translator->trans('Hello World!')."\n";
```

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Tests/Extension/TranslationExtensionTest.php [hunk 1]
```diff
@@ -87,17 +87,17 @@ public function getTransTests()
 
             // transchoice
             array('{% transchoice count from "messages" %}{0} There is no apples|{1} There is one apple|]1,Inf] There is %count% apples{% endtranschoice %}',
-                'There is no apples', array('count' => 0)),
+                'There is no apples', array('count' => 0), ),
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Tests/Extension/TranslationExtensionTest.php]: This looks wrong.

Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Tests/Extension/TranslationExtensionTest.php]: fixed, it is an issue in cs-fixer

**Comment by steffenbrem** [comment]
Would love to see this in the translation component! ❤️ 

@aitboudad Do you still plan on making this ready?

**Comment by lunetics** [comment]
looks good, any updates on that PR

**Comment by aitboudad** [comment]
I wish to accomplish this part but I need some feedback's from the core teams, I don't want to spend much more time like I did before (#16286, #14671, #13986) and at end it's ignored

**Comment by nicolas-grekas** [comment]
@sstok provided a link to this issue, which then triggers this naive question: what is this supposed to solve?
Is it for loading catalogue at runtime with different strategies?
If yes, then I think this may not be a good idea at all, thanks to PHP7.
On PHP7, arrays are stored in shared memory. There is not more efficient strategy, since even for big catalogues, they cost just zero per request.

**Comment by mpdude** [comment]
@nicolas-grekas you mean arrays initialized statically in code are shared, similar to code in the opcache?

**Comment by nicolas-grekas** [comment]
yes, that's what I mean, see https://blog.blackfire.io/php-7-performance-improvements-immutable-arrays.html

**Comment by sstok** [comment]
I will reply tomorrow when my head isn't hurting so bad 🙍‍♂️😉

**Comment by sstok** [comment]
Moving the MessageCatalogue outside of the Translator allows to make it possible to use a Lazy loading MessageCatalogueProvider (eg. ServiceProvider based or Cache decorated) so you don't have to register all the Loaders every time this service is initialized (once per request, but still).

**Comment by aitboudad** [comment]
The main reason is delegating the loading message catalogue (caching, loaders, locale fallback) into separate class which allow us a bunch of things:
- Using the composition over inheritance
- Removing The Translator from FrameworkBundle (less deps)
- Allow disabling fallback catalogues thank to composition (alternative of https://github.com/symfony/symfony/issues/14380 rfc)
- Lazy loading of loaders
- Allow changing the cache layer (support PSR-6 cache?)

I'll try to split this PR in small commits for easy review but before let me know WDYT in overall.

**Comment by Nyholm** [comment]
:+1:

**Comment by nicolas-grekas** [comment]
rebase needed then :)

**Comment by nicolas-grekas** [comment]
@aitboudad still wanting to work on this? Could you please rebase before we enter feat. freeze? (end of week)

**Comment by koemeet** [comment]
gogogoo! finally its happening 🍺

**Comment by fabpot** [comment]
Shall we close? Anyone willing to take over?

**Pull request closed without merging**